### DISCUSSION TITLE
Add an initial implementation of `AppInterceptor`

### DISCUSSION THREAD

**Discussion by luongvo** [nimblehq/android-templates#522] [RFCs]
## Issue

- [`Interceptor`](https://square.github.io/okhttp/features/interceptors/) is an OkHttp component to intercept & modify network requests.
- The template project has a built-in OkHttp & Retrofit configuration for the new project.
- The incorrect order when adding interceptors to OkHttpClient could produce unexpected behaviors. For example, adding `AppInterceptor` after adding `HttpLoggingInterceptor` could make the `HttpLoggingInterceptor` unable to log the full network request body, which contains additional headers in `AppInterceptor` (we faced this issue when leveraging the old codebase for the new working project) ❌ 
	```
	    @Provides
	    fun provideOkHttpClient(
	        apiInterceptor: ApiInterceptor,
	        chuckerInterceptor: ChuckerInterceptor,
	    ) = OkHttpClient.Builder().apply {
	        if (BuildConfig.DEBUG) {
	            addInterceptor(HttpLoggingInterceptor().apply {
	                level = HttpLoggingInterceptor.Level.BODY
	            })
	            addInterceptor(chuckerInterceptor)
	            addInterceptor(apiInterceptor)
	            readTimeout(READ_TIME_OUT, TimeUnit.SECONDS)
	        }
	    }.build()
	```
 
## Solution

- Add an initial implementation of `AppInterceptor`
```
class ApiInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val requestBuilder = chain.request().newBuilder()

		// TODO Inject additional request headers

        return chain.proceed(requestBuilder.build())
    }
}
```

- Add `AppInterceptor` to `OkHttpClient` in the correct order + apply some improvements

```
private const val CONNECTION_TIMEOUT_IN_SECOND = 30L

    @Provides
    fun provideOkHttpClient(
        apiInterceptor: ApiInterceptor,
        chuckerInterceptor: ChuckerInterceptor,
    ) = OkHttpClient.Builder().apply {
        connectTimeout(CONNECTION_TIMEOUT_IN_SECOND, TimeUnit.SECONDS)
        readTimeout(CONNECTION_TIMEOUT_IN_SECOND, TimeUnit.SECONDS)
        writeTimeout(CONNECTION_TIMEOUT_IN_SECOND, TimeUnit.SECONDS)
        
        addInterceptor(apiInterceptor)
        if (BuildConfig.DEBUG) {
            addInterceptor(HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            })
        }
        addInterceptor(chuckerInterceptor)
    }.build()
```

## Who Benefits?

Developers

## What's Next?

- [ ] Add an initial implementation of `AppInterceptor` to the Compose Template.

**Comment by manh-t** [comment]
Voted ✅

**Comment by luongvo** [comment]
cc @nimblehq/chapter-android
